File: ios/Runner/Transfer/PeerTrustStore.swift
import Foundation

/// Trust-on-first-use cache for peer cert fingerprints — Swift mirror of the
/// Android `PeerTrustStore.kt`.
///
/// First time we complete a Karlshare handshake with a peer (TLS auth +
/// fingerprint binding verified), we cache their device UUID → fingerprint.
/// Subsequent connections with the same UUID must present the same
/// fingerprint, else we refuse — that's the signal someone has spoofed the
/// deviceId. Stored in UserDefaults as base64(SHA-256) keyed by UUID.
final class PeerTrustStore {

    enum Result {
        /// Brand-new peer — caller should pin the fingerprint after success.
        case firstSeen
        /// Same peer we've seen before. Safe to proceed.
        case known
        /// Same deviceId but a different fingerprint — refuse.
        case mismatch(expected: String, actual: String)
    }

    private let defaults: UserDefaults
    private let keyPrefix = "karlshare_peer."

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    /// Checks the proposed (deviceId, fingerprint) pair against the cache.
    /// Does NOT write — call `pin` after the handshake completes successfully.
    func verify(deviceId: UUID, fingerprint: Data) -> Result {
        let encoded = fingerprint.base64EncodedString()
        guard let existing = defaults.string(forKey: key(deviceId)) else {
            return .firstSeen
        }
        return existing == encoded
            ? .known
            : .mismatch(expected: existing, actual: encoded)
    }

    func pin(deviceId: UUID, fingerprint: Data) {
        defaults.set(fingerprint.base64EncodedString(), forKey: key(deviceId))
    }

    func forget(deviceId: UUID) {
        defaults.removeObject(forKey: key(deviceId))
    }

    func forgetAll() {
        for k in defaults.dictionaryRepresentation().keys where k.hasPrefix(keyPrefix) {
            defaults.removeObject(forKey: k)
        }
    }

    private func key(_ id: UUID) -> String { keyPrefix + id.uuidString }
}

File: android/app/src/main/kotlin/com/karlshare/karlshare/transfer/PeerTrustStore.kt
package com.karlshare.karlshare.transfer

import android.content.Context
import android.util.Base64
import android.util.Log
import java.util.UUID

/**
 * Trust-on-first-use cache for peer cert fingerprints.
 *
 * First time we successfully complete a Karlshare handshake with a peer
 * (TLS auth + fingerprint binding verified), we cache their device UUID →
 * fingerprint mapping. Subsequent connections with the same UUID must
 * present the same fingerprint, else we refuse the connection — that's the
 * signal that someone has spoofed the deviceId.
 *
 * Stored in SharedPreferences as base64(SHA-256) strings keyed by UUID.
 */
class PeerTrustStore(context: Context) {

    companion object {
        private const val TAG = "PeerTrustStore"
        private const val PREFS = "karlshare_peers"
    }

    private val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    /**
     * Result of consulting the trust store before completing a handshake.
     */
    sealed class Result {
        /** Brand-new peer — caller should pin the fingerprint after success. */
        object FirstSeen : Result()

        /** Same peer we've seen before. Safe to proceed. */
        object Known : Result()

        /** Same deviceId but a different fingerprint — refuse. */
        data class Mismatch(val expected: String, val actual: String) : Result()
    }

    /**
     * Checks the proposed [(deviceId, fingerprint)] pair against the cache.
     * Does NOT write — call [pin] after the handshake completes successfully
     * so we don't pin a fingerprint that turned out to belong to a session
     * we later aborted.
     */
    fun verify(deviceId: UUID, fingerprint: ByteArray): Result {
        val encoded = Base64.encodeToString(fingerprint, Base64.NO_WRAP)
        val existing = prefs.getString(deviceId.toString(), null)
        return when {
            existing == null -> Result.FirstSeen
            existing == encoded -> Result.Known
            else -> {
                Log.w(TAG, "fingerprint mismatch for $deviceId")
                Result.Mismatch(existing, encoded)
            }
        }
    }

    fun pin(deviceId: UUID, fingerprint: ByteArray) {
        val encoded = Base64.encodeToString(fingerprint, Base64.NO_WRAP)
        prefs.edit().putString(deviceId.toString(), encoded).apply()
    }

    fun forget(deviceId: UUID) {
        prefs.edit().remove(deviceId.toString()).apply()
    }

    fun forgetAll() {
        prefs.edit().clear().apply()
    }
}
